### PR TITLE
Use standard dependency workflow

### PR DESCRIPTION
**Pull request by joshuatcasey**
Use standard dependency workflow

### CODE REVIEW

**Comment by joshuatcasey** [comment]
At this point new version retrieval (which includes metadata generation) and dependency testing are complete.

# Retrieval

Note that it's pulling versions from https://www.php.net/releases/active, which is different than the `dep-server` implementation. This URL gives back the most recent releases for each of `7.4`, `8.0`, and `8.1`, which is what we need. Theoretically it's possible for the automation to miss a release if multiple new patches are dropped on a release line very quickly together, but that seems unlikely given that the automation can run hourly and the releases typically occur weekly to monthly.

# Metadata Generation

This step includes ASC signature verification, which did not occur in `dep-server`. I recommend removing this verification.

There are some logistical hurdles in implementing this. Each release line can be signed by one of multiple keys, and there are multiple release lines. I did not copy the public keys into the retrieval source code since they are located on openpgp.org, which is different than where the files are located (php.net). 

For example, I ran the verification step on the following versions and had these results:
* `8.1.12`: verification succeeds
* `8.0.25`: turns out this tarball is signed by fingerprint `2C16C765DBE54A088130F1BC4B9B5F600B55F3B4`, which is not found on `openpgp.org`.
* `7.4.33`: `openpgp: invalid data: entity without any identities` for both public keys

Given these hurdles, I seems unlikely that new versions on these release lines will pass verification.

# Dependency testing

This is almost 90% identical with https://github.com/paketo-buildpacks/dep-server/tree/main/actions/test-dependency/dependency-tests/php, I just added the checksum verification for a little extra confidence.

**Comment by joshuatcasey** [comment]
Note: `libdownloader` is a way to easily download a file. Feel free to refactor it away, inline its code, fork it to `paketo-community`, etc etc.

**Comment by fg-j** [comment]
# Compilation
## What's done
I've added `dependency/actions/compile`, which currently includes compilation Dockerfiles for bionic and jammy dependencies. These share an entrypoint script. As of now, the compilation runs as expected for PHP version lines `7.4` `8.0` and `8.1`. Note that only PHP 8.1 is compatible with Jammy Jellyfish, so the dependency automation is only set up to build PHP 8.1 for Jammy. The logic that compiles PHP is lifted directly from https://github.com/cloudfoundry/binary-builder and https://github.com/cloudfoundry/binary-builder/tree/main/cflinuxfs4, with some tweaks:
- I've removed `*.rb` files that don't relate to compiling PHP
- I've tweaked the various `cp ` steps that run during compilation to add the `-v` verbose flag, for more visibility into what's going on
- I've shifted OS package installation into the Dockerfiles

Note that  `cflinuxfs4` is a fork of its parent written to be compatible with Ruby 3, which is needed to run on Jammy.

The outputs of compilation for version lines `7.4`, `8.0`, and `8.1` pass the tests in `dependency/test`! This is a good indicator that these PHP dependencies just might work in the buildpack.

## What's next
- [ ] validate that compilation outputs actually work as expected in the buildpack
- [ ] clean up more dead code in `binary-builder`
- [ ] resolve the key signing issue that keeps cropping up during retrieval, as Josh mentioned above
- [ ] (eventually) extract PHP compilation from ruby code and/or refactor ruby code

**Comment by sophiewigmore** [comment]
@fg-j thank you for the thorough update & work on this!! For future context, can you briefly explain the main issue(s) you ran into trying to rewrite the compilation code (before you decided to pull it in as-is).

**Comment by fg-j** [comment]
The primary reason I ruled re-writing the compilation code out of scope is that the it relies on an external package called [MiniPortile](https://www.rubydoc.info/gems/mini_portile2/2.7.1) to handle the details of actually compiling PHP and each of the extensions that are added to Paketo's PHP distro. MiniPortile defines the `recipe` abstraction that is used in the compilation code. This MiniPortile type defines default methods like `cook` and `activate`, from which the PHP-related classes inherit. See [this definition of `PHPRecipe`](https://github.com/cloudfoundry/binary-builder/blob/69a6515f1ac1bbf89af51fb3e012e9e0d5181af6/recipe/php_recipe.rb#L4) and [this definition of `BaseRecipe`](https://github.com/cloudfoundry/binary-builder/blob/69a6515f1ac1bbf89af51fb3e012e9e0d5181af6/recipe/base.rb#L8).

To move away from the Ruby code and its abstractions will require 1) understanding fully what MiniPortile is doing under the hood, 2) deciding whether to use a dependency-compiling abstraction in another language, 3) developing some strong integration tests that will provide confidence during a refactor, 4) translating from `MiniPortile` to the other abstraction, or rewriting the logic within `MiniPortile` from scratch.

I can see why using an external abstraction for compilation is useful here – many of the same steps are required for each of PHP's many extensions. But the "automagic" of the current design also makes it hard to reverse-engineer with confidence. 

In summary, I think reworking PHP compilation is worth doing in the long term, but it's nontrivial. I think moving the old logic into the new federated dependency model is a decent short term win, since we get better feedback loops and cleaner containerization/portability.

**Comment by TisVictress** [comment]
Closing this draft since PR https://github.com/paketo-buildpacks/php-dist/pull/469 is open.
